budgetportal/prov_infra_project/charts.py

```diff
@@ -0,0 +1,131 @@
+        quarter_is_found = False
+        date, quarter_label, fin_year_label = extract_date_quarter_year(i, fin_year)
+        total_spent_to_date, total_spent_in_quarter = compute_total_spent(snapshot, i)
+        for data in chart_data["snapshots"]:
+            if data["date"] == date:
+                data.update(
+                    {
+                        "total_spent_to_date": total_spent_to_date,
+                        "total_spent_in_quarter": total_spent_in_quarter,
+                    }
+                )
+                quarter_is_found = True
+
+        if quarter_is_found is False:
+            chart_data["snapshots"].append(
+                {
+                    "date": date,
+                    "quarter_label": quarter_label,
+                    "financial_year_label": fin_year_label,
+                    "total_spent_to_date": total_spent_to_date,
+                    "total_spent_in_quarter": total_spent_in_quarter,
+                    "total_estimated_project_cost": None,
+                    "status": None,
+                }
+            )
+    return chart_data
+
+
+def extract_date_quarter_year(quarter_number, financial_year):
+    dates = {1: "-06-30", 2: "-09-30", 3: "-12-31", 4: "-03-31"}
+    if quarter_number == 4:
+        date = str(int(financial_year.get_starting_year()) + 1) + dates[quarter_number]
+    else:
+        date = financial_year.get_starting_year() + dates[quarter_number]
+
+    quarter_label = "END Q{}".format(quarter_number)
+    financial_year_label = ""
+    if quarter_number == 1:
+        financial_year_label = financial_year.slug
+
+    return date, quarter_label, financial_year_label
+
+
+def compute_total_spent(project_snapshot, quarter_number):
+    total_spent_in_quarter = None
+    total_spent_to_date = None
+    total_from_previous_years = project_snapshot.expenditure_from_previous_years_total
+    if total_from_previous_years is not None:
+        total_spent_to_date = total_from_previous_years
+
+    for i in range(1, quarter_number + 1):
+        field = "actual_expenditure_q{}".format(i)
+        quarterly_spent = getattr(project_snapshot, field)
+        if quarterly_spent is None:
+            total_spent_to_date = None
+
+        if total_spent_to_date:
+            total_spent_to_date += quarterly_spent
+
+        if i == quarter_number:
+            total_spent_in_quarter = quarterly_spent
+
+    return total_spent_to_date, total_spent_in_quarter
+
+
+def find_latest_snapshot_with_dates(project_snapshots):
+    for snapshot in project_snapshots:
+        if (
+            snapshot.estimated_construction_start_date
+            and snapshot.estimated_construction_end_date
+        ):
+            return snapshot
+
+    return None
+
+
+def extract_events(project_snapshot):
+    events = []
+    start_date = project_snapshot.estimated_construction_start_date
+    events.append(
+        {"date": str(start_date), "label": "Estimated construction start date"}
```

Comment:
> - All the date fields in the snapshots for which the latest snapshot of all snapshots is not null should be output as elements of the events array using the column name in the excel files passed to IRMSnapshot as their label.
>   - hardcode the field names and labels - don't try and introspect and generate these based on the type of field - I'm just using the fact that they're all currently date fields to simplify this description.

```
    start_date = models.DateField(blank=True, null=True)
    estimated_construction_start_date = models.DateField(blank=True, null=True)
    estimated_completion_date = models.DateField(blank=True, null=True)
    contracted_construction_end_date = models.DateField(blank=True, null=True)
    estimated_construction_end_date = models.DateField(blank=True, null=True)
```

Reply:
Please use the label `Project Start Date` for  `start_date`

Reply:
All date fields added to extract_events

https://github.com/vulekamali/datamanager/pull/215/commits/8c598af2001e62b60960b89ddf88e9461cc19e5a